apisix/discovery/consul/init.lua

```diff
@@ -175,81 +188,291 @@ local function get_retry_delay(retry_delay)
+            and catalog_res.status)
+    if catalog_error_info then
+        log.error("connect consul: ", consul_server.consul_server_url,
+            " by sub url: ", consul_server.consul_watch_catalog_url,
+            ", got catalog result: ", json_delay_encode(catalog_res, true),
+            ", with error: ", catalog_error_info)
 
-    log.info("consul_server: ", json_delay_encode(consul_server, true))
-    local watch_result, watch_err = consul_client:get(consul_server.consul_watch_sub_url)
-    local watch_error_info = (watch_err ~= nil and watch_err)
-            or ((watch_result ~= nil and watch_result.status ~= 200)
-            and watch_result.status)
-    if watch_error_info then
+        retry_delay = get_retry_delay(retry_delay)
+        log.warn("get all svcs got err, retry connecting consul after ", retry_delay, " seconds")
+        core_sleep(retry_delay)
+
+        check_keepalive(consul_server, retry_delay)
+        return
+    end
+
+    -- get health index
+    local success, health_res, health_err = pcall(function()
+        return consul_client:get(consul_server.consul_watch_health_url)
+    end)
+    if not success then
         log.error("connect consul: ", consul_server.consul_server_url,
-            " by sub url: ", consul_server.consul_watch_sub_url,
-            ", got watch result: ", json_delay_encode(watch_result, true),
-            ", with error: ", watch_error_info)
+            " by sub url: ", consul_server.consul_watch_health_url,
+            ", got health result: ", json_delay_encode(health_res, true))
+        check_keepalive(consul_server, retry_delay)
+        return
+    end
+    local health_error_info = (health_err ~= nil and health_err)
+            or ((health_res ~= nil and health_res.status ~= 200)
+            and health_res.status)
+    if health_error_info then
+        log.error("connect consul: ", consul_server.consul_server_url,
+            " by sub url: ", consul_server.consul_watch_health_url,
+            ", got health result: ", json_delay_encode(health_res, true),
+            ", with error: ", health_error_info)
 
         retry_delay = get_retry_delay(retry_delay)
-        log.warn("retry connecting consul after ", retry_delay, " seconds")
+        log.warn("get all svcs got err, retry connecting consul after ", retry_delay, " seconds")
         core_sleep(retry_delay)
 
-        goto ERR
+        goto ERROR
     end
 
     log.info("connect consul: ", consul_server.consul_server_url,
-        ", watch_result status: ", watch_result.status,
-        ", watch_result.headers.index: ", watch_result.headers['X-Consul-Index'],
+        ", catalog_result status: ", catalog_res.status,
+        ", catalog_result.headers.index: ", catalog_res.headers['X-Consul-Index'],
         ", consul_server.index: ", consul_server.index,
         ", consul_server: ", json_delay_encode(consul_server, true))
 
     -- if current index different last index then update service
```

Comment:
if the current index is different from the last index, then update the service

Reply:
done